packages/next-on-pages/templates/_worker.js/utils/matcher.ts

```diff
@@ -13,34 +17,64 @@ type HasFieldRequestProperties = {
  */
 export function hasField(
 	has: VercelHasField,
-	{ url, cookies, headers }: HasFieldRequestProperties,
-): boolean {
+	{ url, cookies, headers, routeDest }: HasFieldRequestProperties,
+): { valid: boolean; newRouteDest?: string } {
 	switch (has.type) {
 		case 'host': {
-			return url.hostname === has.value;
+			return { valid: url.hostname === has.value };
 		}
 		case 'header': {
 			if (has.value !== undefined) {
-				return !!headers.get(has.key)?.match(has.value);
+				const match = matchPCRE(has.value, headers.get(has.key));
+				return {
+					valid: !!match.match,
+					newRouteDest: tryApplyMatch(routeDest, match),
+				};
```

Comment:
The current implementation seems to always have this `matchPCRE` + return object code, so there is a bit of code duplication, I've tried to remove the code duplication and ended up with something like this:
```diff
diff --git a/packages/next-on-pages/templates/_worker.js/utils/matcher.ts b/packages/next-on-pages/templates/_worker.js/utils/matcher.ts
index 666ae895..f1e40aed 100644
--- a/packages/next-on-pages/templates/_worker.js/utils/matcher.ts
+++ b/packages/next-on-pages/templates/_worker.js/utils/matcher.ts
@@ -1,4 +1,3 @@
-import type { MatchPCREResult } from './pcre';
 import { applyPCREMatches, matchPCRE } from './pcre';
 
 type HasFieldRequestProperties = {
@@ -17,7 +16,7 @@ type HasFieldRequestProperties = {
  */
 export function hasField(
        has: VercelHasField,
-       { url, cookies, headers, routeDest }: HasFieldRequestProperties,
+       { url, cookies, headers, routeDest }: HasFieldRequestProperties
 ): { valid: boolean; newRouteDest?: string } {
        switch (has.type) {
                case 'host': {
@@ -25,11 +24,12 @@ export function hasField(
                }
                case 'header': {
                        if (has.value !== undefined) {
-                               const match = matchPCRE(has.value, headers.get(has.key));
-                               return {
-                                       valid: !!match.match,
-                                       newRouteDest: tryApplyMatch(routeDest, match),
-                               };
+                               return getHasFieldPCREMatchResult(
+                                       has.value,
+                                       routeDest,
+                                       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                                       headers.get(has.key)!
+                               );
                        }
 
                        return { valid: headers.has(has.key) };
@@ -38,22 +38,18 @@ export function hasField(
                        const cookie = cookies[has.key];
 
                        if (cookie && has.value !== undefined) {
-                               const match = matchPCRE(has.value, cookie);
-                               return {
-                                       valid: !!match.match,
-                                       newRouteDest: tryApplyMatch(routeDest, match),
-                               };
+                               return getHasFieldPCREMatchResult(has.value, routeDest, cookie);
                        }
 
                        return { valid: cookie !== undefined };
                }
                case 'query': {
                        if (has.value !== undefined) {
-                               const match = matchPCRE(has.value, url.searchParams.get(has.key));
-                               return {
-                                       valid: !!match.match,
-                                       newRouteDest: tryApplyMatch(routeDest, match),
-                               };
+                               return getHasFieldPCREMatchResult(
+                                       has.value,
+                                       routeDest,
+                                       url.searchParams.get(has.key) ?? ''
+                               );
                        }
 
                        return { valid: url.searchParams.has(has.key) };
@@ -62,19 +58,33 @@ export function hasField(
 }
 
 /**
- * Try to apply a PCRE match's named capture groups to a destination.
+ * Gets the has field result from a PCRE match
  *
- * @param dest Destination to apply match to.
- * @param match Matched PCRE result.
- * @returns The destination with the match applied, or `undefined` if there was no match.
+ * @param hasValue the value of the has field
+ * @param routeDest original route destination
+ * @param target target string to match
+ * @returns an object indicating if the match is valid and the new route destination if existent
  */
-function tryApplyMatch(
-       dest: string | undefined,
-       { match, captureGroupKeys }: MatchPCREResult,
-): string | undefined {
-       if (dest && match && captureGroupKeys.length) {
-               return applyPCREMatches(dest, match, captureGroupKeys, { namedOnly: true });
+function getHasFieldPCREMatchResult(
+       hasValue: string,
+       routeDest: string | undefined,
+       target: string
+): ReturnType<typeof hasField> {
+       const match = matchPCRE(hasValue, target);
+       let newRouteDest: string | undefined = undefined;
+
+       // if there are capture groups apply them to get a new route destination
+       if (routeDest && match.match && match.captureGroupKeys.length) {
+               newRouteDest = applyPCREMatches(
+                       routeDest,
+                       match.match,
+                       match.captureGroupKeys,
+                       { namedOnly: true }
+               );
        }
 
-       return undefined;
+       return {
+               valid: !!match.match,
+               newRouteDest,
+       };
 }
```

I __think__ something along these lines might be nicer but I am not completely sure, what do you think?

If you prefer the current implementation I'm totally fine to go with it as well 🙂

Reply:
I could have bet money on you replying this!! I aaallllllmmmmoooossstttt made that change earlier this week in anticipation that you might say that. 😅

Reply:
sorry for being too predictable 🙈 

anyways as I said it's just a suggestion, I'm totally ok with your solution too

my suggestion also isn't any shorter (I hoped it would be before writing it) so I don't think it's objectively better